### PR TITLE
chore (perf): lazy load NavHistoryService in UserDomainAffinityProvider

### PR DESCRIPTION
**Pull request by rlr**
r? @Mardak

### CODE REVIEW

File: system-addon/lib/UserDomainAffinityProvider.jsm [hunk 1]
```diff
@@ -91,6 +89,13 @@ this.UserDomainAffinityProvider = class UserDomainAffinityProvider {
     }
   }
 
+  get history() {
+    if (!this._history) {
+      this._history = Cc["@mozilla.org/browser/nav-history-service;1"].getService(Ci.nsINavHistoryService);
```
Comment (Mardak) [inline system-addon/lib/UserDomainAffinityProvider.jsm]: There's a `XPCOMUtils.defineLazyServiceGetter` although unit-entry just drops it so individual tests manually set a global matching the jsm's expected global.

We could fix up unit-entry to do something smarter.. but maybe later.. :p

**Comment by rlr** [comment]
This better?

```
diff --git a/system-addon/lib/UserDomainAffinityProvider.jsm b/system-addon/lib/UserDomainAffinityProvider.jsm
index 2fb23761..88cbebbc 100644
--- a/system-addon/lib/UserDomainAffinityProvider.jsm
+++ b/system-addon/lib/UserDomainAffinityProvider.jsm
@@ -4,8 +4,9 @@
 "use strict";

 ChromeUtils.import("resource://gre/modules/Services.jsm");
+ChromeUtils.import("resource://gre/modules/XPCOMUtils.jsm");

-const history = Cc["@mozilla.org/browser/nav-history-service;1"].getService(Ci.nsINavHistoryService);
+XPCOMUtils.defineLazyServiceGetter(this, "NavHistoryService", "@mozilla.org/browser/nav-history-service;1", "nsINavHistoryService");

 const DEFAULT_TIME_SEGMENTS = [
   {"id": "hour", "startTime": 3600, "endTime": 0, "weightPosition": 1},
@@ -243,7 +244,7 @@ this.UserDomainAffinityProvider = class UserDomainAffinityProvider {
    */
   queryVisits(ts) {
     const visitCounts = {};
-    const query = history.getNewQuery();
+    const query = NavHistoryService.getNewQuery();
     const wwwRegEx = /^www\./;

     query.beginTimeReference = query.TIME_RELATIVE_NOW;
@@ -252,11 +253,11 @@ this.UserDomainAffinityProvider = class UserDomainAffinityProvider {
     query.endTimeReference = query.TIME_RELATIVE_NOW;
     query.endTime = (ts.endTime && ts.endTime !== 0) ? -(ts.endTime * 1000 * 1000) : 0;

-    const options = history.getNewQueryOptions();
+    const options = NavHistoryService.getNewQueryOptions();
     options.sortingMode = options.SORT_BY_VISITCOUNT_DESCENDING;
     options.maxResults = this.maxHistoryQueryResults;

-    const {root} = history.executeQuery(query, options);
+    const {root} = NavHistoryService.executeQuery(query, options);
     root.containerOpen = true;
     for (let i = 0; i < root.childCount; i++) {
       let node = root.getChild(i);
diff --git a/system-addon/test/unit/lib/UserDomainAffinityProvider.test.js b/system-addon/test/unit/lib/UserDomainAffinityProvider.test.js
index 8ac08617..c8f53957 100644
--- a/system-addon/test/unit/lib/UserDomainAffinityProvider.test.js
+++ b/system-addon/test/unit/lib/UserDomainAffinityProvider.test.js
@@ -43,6 +43,7 @@ describe("User Domain Affinity Provider", () => {
     globals.sandbox.stub(history, "executeQuery").returns({root: {childCount: 1, getChild: index => ({uri: testUrl, accessCount: 1})}});
     globals.sandbox.stub(history, "getNewQuery").returns({"TIME_RELATIVE_NOW": 1});
     globals.sandbox.stub(history, "getNewQueryOptions").returns({});
+    globals.set("NavHistoryService", history);

     instance = new UserDomainAffinityProvider(TIME_SEGMENTS, PARAMETER_SETS);
   });
```

Doesn't make unit-entry smarter yet though. I'll try that next.

**Comment by Mardak** [comment]
I suppose if we're okay with `history` being a top level global object in the jsm, then `XPCOMUtils.defineLazyServiceGetter(this, "history", …)` would reduce the number of changes ;)

**Comment by rlr** [comment]
I guess there is already a `history` in the global scope so the overrider won't let me override 😠 

Also, making `defineLazyServiceGetter` smarter in unit-entry seems hard because that runs at import time. Might punt on that 😝

**Comment by rlr** [comment]
or maybe I'll try to use https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy

**Comment by rlr** [comment]
btw, I tried this but no workey:

```
-    defineLazyServiceGetter() {},
+    defineLazyServiceGetter(_1, _2, _3, _4) {
+      TEST_GLOBAL[_2] = _3 in TEST_GLOBAL ? TEST_GLOBAL.Cc[_3].getService() : {};
+    },
```

File: system-addon/lib/UserDomainAffinityProvider.jsm [hunk 2]
```diff
@@ -4,8 +4,9 @@
 "use strict";
 
 ChromeUtils.import("resource://gre/modules/Services.jsm");
+ChromeUtils.import("resource://gre/modules/XPCOMUtils.jsm");
 
-const history = Cc["@mozilla.org/browser/nav-history-service;1"].getService(Ci.nsINavHistoryService);
+XPCOMUtils.defineLazyServiceGetter(this, "NavHistoryService", "@mozilla.org/browser/nav-history-service;1", "nsINavHistoryService");
```
Comment (Mardak) [inline system-addon/lib/UserDomainAffinityProvider.jsm]: Oh, alternatively `ChromeUtils.defineModuleGetter(this, "PlacesUtils", "resource://gre/modules/PlacesUtils.jsm");` then access `PlacesUtils.history`?

**Comment by as-pine-proxy** [comment]
[Treeherder: d5e4217996510616a69857da7364014962679f67](https://treeherder.mozilla.org/#/jobs?repo=pine&revision=d5e4217996510616a69857da7364014962679f67)

**Pull request merged**